new/supermarket-release-channels.md

```diff
@@ -0,0 +1,93 @@
+---
+RFC: unassigned
+Author: Daniel DeLeo <dan@chef.io>
+Status: Draft
+Type: Standards Track
+---
+
+# Supermarket Release Channels
+
+Add the ability to assign supermarket artifacts (cookbook versions) to a named
+channel, and to filter the universe endpoint data by specifying a channel name.
+
+## Motivation
+
+    As a Supermarket User,
+    I want to promote my artifacts through named channels,
+    so that I can release them according to my testing and release policy.
+
+Many users implement a testing pipeline to automate testing and release of chef
+artifacts. An example of this kind of system is the workflow component of the
+Chef Automate product. Frequently it is necessary or convenient to publish an
+artifact to a Supermarket service at an early pipeline stage, so that the
+artifact can be distributed to other systems for testing. For example, when
+using Chef Automate, the default behavior is to publish the artifact in the
+build phase, so that it may be retrieved by other systems during acceptance
+testing. In the current implementation, publishing the artifact makes it
+available to all users of the Supermarket service. For example, a berkshelf
+user developing a cookbook may retrieve this artifact when installing
+dependencies. This is undesireable because the artifact has not yet passed
+acceptance testing and therefore may contain flaws. Additionally, in the case
+of Chef Automate, there is a manual release gate that occurs later in the
+pipeline; it would be preferable that users only access artifacts that have
+passed this gate.
+
+In other artifact storage systems, artifacts may be assigned to channels and
+promoted as they pass automated or manual verification. For example, each Chef
+Client build is initially created in an "unstable" channel for distribution to
+automated testing systems. If the package passes automated testing, it is
+promoted to the "current" channel where users may access it.
+
+Adding release channels gives Supermarket users this same ability to control
+the release process for their Chef artifacts.
+
+## Specification
+
+#### Association of Artifacts to Channels
+
+Artifacts may belong to an arbitrary number of channels. We expect that in the
+case of Chef Automate, users will want artifacts to remain available in the
+union channel even after promoting that artifact to rehearsal.
+
+#### Artifact Versions and Channels
+
+Channels may have any number of versions of an artifact, so that old versions
+remain available after a newer version is promoted to that channel.
+
+There are some use cases where it is desirable to only allow one version of an
+artifact in a channel. Users of Chef Automate, for example, are encouraged to
+only use one version of an artifact in a given environment. Thefore, the
+user may specify that an artifact (TODO: is this config on the channel, or an
```

Comment:
Exclusivity seems weird to me. I understand the idea behind it, but I don't know of an existing artifact storage system that implements this. I am familiar with the idea that "you get the latest version of the requested artifact unless you specify a version constraint".

Thinking more about it, I can understand the use case in a private supermarket. I don't have enough context around private supermarket use cases to answer this TODO. But the simplest way to support this feature seems to be to specify exclusivity on a channel level, so thats what I would vote for if we do not know of another customer use case.